cpp/dolfinx/fem/DirichletBC.h

```diff
@@ -495,76 +508,82 @@ class DirichletBC
   void set(std::span<T> x, std::optional<std::span<const T>> x0,
            T alpha = 1) const
   {
-    std::int32_t x_size = x.size();
-    if (alpha == T(0)) // Optimisation for when alpha == 0
+    // set_fn is a lambda which gets evaluated for every index in [0,
+    // _dofs0.size()) and its result is assigned to x[_dofs0[i]].
+    auto apply = [&](std::invocable<std::int32_t> auto set_fn)
     {
-      for (std::int32_t idx : _dofs0)
+      static_assert(
+          std::is_same_v<std::invoke_result_t<decltype(set_fn), std::int32_t>,
+                         T>);
+
+      std::int32_t x_size = x.size();
+      for (std::int32_t i = 0; i < _dofs0.size(); ++i)
```

Comment:
Surprised this compiles with our strict flags - `_dofs0.size()` has type `std::size_t`.